xclim/sdba/base.py

```diff
@@ -366,23 +403,198 @@ def parse_group(func: Callable) -> Callable:
+
+            if uses_dask(ds):
+                # Use dask if any of the input is dask-backed.
+                chunks = (
+                    dict(ds.chunks)
+                    if isinstance(ds, xr.Dataset)
+                    else dict(zip(ds.dims, ds.chunks))
+                )
+                if (
+                    group is not None
+                    and group.dim in chunks
+                    and len(chunks[group.dim]) > 1
+                ):
+                    raise ValueError(
+                        f"The dimension over which we group cannot be chunked ({group.dim} has chunks {chunks[group.dim]})."
+                    )
+            else:
+                chunks = None
+
+            # Make translation dict
+            if group is not None:
+                placeholders = {"<PROP>": group.prop, "<DIM>": group.dim}
+            else:
+                placeholders = {}
+
+            # Get new dimensions (in order), translating placeholders to real names.
+            new_dims = [placeholders.get(dim, dim) for dim in out_dims]
+            reduced_dims = [placeholders.get(dim, dim) for dim in red_dims]
+
+            for dim in new_dims:
+                if dim in ds.dims and dim not in reduced_dims:
+                    raise ValueError(
+                        f"Dimension {dim} is meant to be added by the computation but it is already on one of the inputs."
+                    )
+
+            # Dimensions untouched by the function.
+            base_dims = list(set(ds.dims) - set(new_dims) - set(reduced_dims))
+
+            # All dimensions of the output data, new_dims are added at the end on purpose.
+            all_dims = base_dims + new_dims
+            # The coordinates of the output data.
+            coords = {}
+            for dim in all_dims:
+                if dim == group.prop:
+                    coords[group.prop] = group.get_coordinate(ds=ds)
+                elif dim == group.dim:
+                    coords[group.dim] = ds[group.dim]
+                elif dim in ds.coords:
+                    coords[dim] = ds[dim]
+                elif dim in kwargs:
+                    coords[dim] = xr.DataArray(kwargs[dim], dims=(dim,), name=dim)
+                else:
+                    raise ValueError(
+                        f"This function adds the {dim} dimension, its coordinate must be provided as a keyword argument."
+                    )
+            sizes = {name: crd.size for name, crd in coords.items()}
+
+            # Create the output dataset, but empty
+            tmpl = xr.Dataset(coords=coords)
+            for var, dims in outvars.items():
+                # Out variables must have the base dims + new_dims
+                dims = base_dims + [placeholders.get(dim, dim) for dim in dims]
+                # duck empty calls dask if chunks is not None
+                tmpl[var] = duck_empty(dims, sizes, chunks)
+
+            def _transpose_on_exit(dsblock, **kwargs):
```

Comment:
This is where the actual work is done right ? The function name suggests it's doing a reshape operation.

Reply:
Isn't it surprising that we have to deal with dimension order within xarray ?

Reply:
True, will give a clearer name.

Yes, a bit surprising. Raised the issue at pydata/xarray#5326.

Reply:
Issue on xarray closed. Conclusion:
As we pass a `template` that is supposed to look like the result, it is expected (but not checked)  that the result fits the template perfectly : sames sizes, same variables and _same dimension order_. We could also try to guess the order, but I think it is simpler to transpose the result. The performance impact should be quite low.